sklearn/linear_model/huber.py

```diff
@@ -0,0 +1,349 @@
+from ..utils import check_consistent_length
+from ..utils import column_or_1d
+from ..utils import safe_mask
+from ..utils.extmath import safe_sparse_dot
+
+
+def _dia_matrix(X):
+    """
+    Old versions of scipy cannot create an empty dia_matrix
+    """
+    if len(X) == 0:
+        return 1
+    return sparse.dia_matrix(X)
+
+
+def _axis0_safe_slice(X, mask, len_mask):
+    """
+    This mask is safer than safe_mask since it returns an
+    empty array, when a sparse matrix is sliced with a boolean mask
+    with all False, instead of raising an unhelpful error in older
+    versions of SciPy.
+
+    See: https://github.com/scipy/scipy/issues/5361
+
+    Also note that we can avoid doing the dot product by checking if
+    the len_mask is not zero in _huber_loss_and_gradient but this
+    is not going to be the bottleneck, since the number of outliers
+    and non_outliers are typically non-zero and it makes the code
+    tougher to follow.
+    """
+    if len_mask != 0:
+        return X[safe_mask(X, mask), :]
+    return np.zeros(shape=(0, X.shape[1]))
+
+
+def _huber_loss_and_gradient(w, X, y, epsilon, alpha, sample_weight=None):
+    """Returns the Huber loss and the gradient.
+
+    Parameters
+    ----------
+    w: ndarray, shape (n_features + 1,) or (n_features + 2,)
+        Feature vector.
+        w[:n_features] gives the feature vector
+        w[-1] gives the scale factor and if the intercept is fit w[-2]
+        gives the intercept factor.
+
+    X: ndarray, shape (n_samples, n_features)
+        Input data.
+
+    y: ndarray, shape (n_samples,)
+        Target vector.
+
+    epsilon: float
+        Measures the robustness of the Huber estimator.
+
+    alpha: float
+        Regularization parameter.
+
+    sample_weight: ndarray, shape (n_samples,), optional
+        Weight assigned to each sample.
+
+    Returns
+    -------
+    loss: float
+        Huber loss.
+
+    gradient: ndarray, shape (n_features + 1,) or (n_features + 2,)
+        Returns the derivative of the Huber loss with respect to each
+        coefficient, intercept and the scale as a vector.
+    """
+    X_is_sparse = sparse.issparse(X)
+    n_samples, n_features = X.shape
+    fit_intercept = (n_features + 2 == w.shape[0])
+    if fit_intercept:
+        intercept = w[-2]
+    sigma = w[-1]
+    w = w[:n_features]
+
+    if sample_weight is not None:
+        n_samples = np.sum(sample_weight)
+
+    # Calculate the values where |y - X'w -c / sigma| > epsilon
+    # The values above this threshold are outliers.
+    linear_loss = y - safe_sparse_dot(X, w)
+    if fit_intercept:
+        linear_loss -= intercept
+    abs_linear_loss = np.abs(linear_loss)
+    outliers_mask = (abs_linear_loss / sigma) > epsilon
+
+    # Calculate the linear loss due to the outliers.
+    # This is equal to (2 * M * |y - X'w -c / sigma| - M**2) * sigma
+    outliers = abs_linear_loss[outliers_mask]
+    num_outliers = np.count_nonzero(outliers)
```

Comment:
I don't think it's great if there is ``num_outliers`` and ``n_outliers`` and they have quite different semantics (which are not explained). Actually I'm not very sure what happens here. ``num_outliers`` is the number of outliers with non-zero loss, right? But the loss is float. How can it be zero?